rs/artifact_pool/src/ingress_pool/peer_counter.rs

```diff
@@ -0,0 +1,185 @@
+use std::{
+    collections::{btree_map::Entry, BTreeMap},
+    ops::Add,
+};
+
+use ic_interfaces::ingress_pool::IngressPoolObject;
+use ic_logger::{warn, ReplicaLogger};
+use ic_types::{CountBytes, NodeId};
+
+pub(super) struct Counter {
+    pub(super) bytes: usize,
+    pub(super) messages: usize,
+}
+
+impl Add for Counter {
+    type Output = Counter;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        Counter {
+            bytes: self.bytes + rhs.bytes,
+            messages: self.messages + rhs.messages,
+        }
+    }
+}
+
+#[derive(Clone)]
+/// Keeps track of the number of messages and total size of the messages per peer.
+pub(super) struct PeerCounters {
+    bytes_counters: PeerCounter,
+    message_counters: PeerCounter,
+}
+
+impl PeerCounters {
+    /// Create a new [`PeerCounters`] without any limits on the number/size of messages.
+    pub(super) fn new(log: ReplicaLogger) -> Self {
+        Self {
+            bytes_counters: PeerCounter::new(log.clone()),
+            message_counters: PeerCounter::new(log),
+        }
+    }
+
+    pub(super) fn observe(&mut self, ingress_message: &IngressPoolObject) {
+        self.message_counters.add(ingress_message.peer_id, 1);
+        self.bytes_counters
+            .add(ingress_message.peer_id, ingress_message.count_bytes());
+    }
+
+    pub(super) fn forget(&mut self, ingress_message: &IngressPoolObject) {
+        self.message_counters.subtract(ingress_message.peer_id, 1);
+        self.bytes_counters
+            .subtract(ingress_message.peer_id, ingress_message.count_bytes());
+    }
+
+    pub(super) fn count_total_bytes(&self) -> usize {
+        self.bytes_counters.total()
+    }
+
+    pub(super) fn get_counters(&self, peer_id: &NodeId) -> Counter {
+        Counter {
+            bytes: self
+                .bytes_counters
+                .counter_per_peer
+                .get(peer_id)
+                .copied()
+                .unwrap_or_default(),
+            messages: self
+                .message_counters
+                .counter_per_peer
+                .get(peer_id)
+                .copied()
+                .unwrap_or_default(),
+        }
+    }
+}
+
+#[derive(Clone)]
+struct PeerCounter {
```

Comment:
nit: the name of the struct is a bit misleading because you have a hashmap of all peers and it is not representative of a single peer. no action, we can fix it later.

Reply:
Yeah, I don't like it either, I've left a TODO and will rename it in a subsequent PR